VDA-SBT/contracts/SoulboundNFT.sol

```diff
@@ -0,0 +1,316 @@
+    }
+
+    /**
+     * @dev Overrides of super class function. Used to emit SBT events.
+     * See {IERC721Upgradeable}
+     */
+    function _afterTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal override virtual {
+        if (to == address(0x0)) {
+            emit Unlocked(tokenId);
+        } else if (to != address(0xdead)) {
+            emit Locked(tokenId);
+        }
+    }
+
+    /**
+     * @dev See {IERC5192}
+     */
+    function locked(uint256 tokenId) external view override returns(bool) {
+        address owner = ERC721Upgradeable.ownerOf(tokenId);
+        if (owner == address(0x0)) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function totalSupply() external view onlyOwner override returns(uint){
+        return _tokenIdCounter.current();
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function addCompanyAccount(address account) external onlyOwner override {
+        require(!_companyAccounts.contains(account), "Existing account");
+        _companyAccounts.add(account);
+
+        emit AddCompanyAccount(account);
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function removeCompanyAccount(address account) external onlyOwner override {
+        require(_companyAccounts.contains(account), "Invalid account");
+        _companyAccounts.remove(account);
+
+        emit RemoveCompanyAccount(account);
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function listCompanyAccounts() external view override returns(address[] memory) {
+        uint length = _companyAccounts.length();
+        address[] memory list = new address[](length);
+        for (uint i = 0; i < length; i++) {
+            list[i] = _companyAccounts.at(i);
+        }
+
+        return list;
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function addSBTType(string calldata sbtType) external onlyOwner override {
+        require(!_sbtTypes.contains(sbtType), "Existing SBT type");
+        _sbtTypes.add(sbtType);
+
+        emit AddSBTType(sbtType);
+    }
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    /*
+    function removeSBTType(string calldata sbtType) external onlyOwner override {
+        require(_sbtTypes.contains(sbtType), "Not registered type");
+        _sbtTypes.remove(sbtType);
+
+        emit RemoveSBTType(sbtType);
+    }
+    */
+
+    /**
+     * @dev See {ISoulboundNFT}
+     */
+    function allowedSBTTypes() external view override returns(string[] memory) {
+        // To-do
+        uint length = _sbtTypes.length();
+
+        string[] memory sbtTypeList = new string[](length);
+
+        for (uint i = 0; i < length; i++) {
+            sbtTypeList[i] = _sbtTypes.at(i);
+        }
+
+        return sbtTypeList;
+    }
+
+    /**
+     * @dev See {IsoulboundNFT}
+     */
+    function claimSBT(
```

Comment:
The caller of this method may be the metadata server, so we need to support the caller specifying the address that should receive the minted NFT